server/src/main/java/org/opensearch/cluster/routing/allocation/decider/WarmDiskThresholdDecider.java

```diff
@@ -0,0 +1,322 @@
+        ByteSizeValue bytesThreshold = diskThresholdSettings.getFreeBytesThresholdLow();
+        if (bytesThreshold != null && bytesThreshold.getBytes() > 0) {
+            return bytesThreshold.getBytes() * (long) dataToFileCacheSizeRatio;
+        }
+
+        // Default fallback
+        return 0;
+    }
+
+    private long calculateFreeSpaceHighThreshold(DiskThresholdSettings diskThresholdSettings, long totalAddressableRemoteSize) {
+        // Check for percentage-based threshold
+        double percentageThreshold = diskThresholdSettings.getFreeDiskThresholdHigh();
+        if (percentageThreshold > 0) {
+            return (long) (totalAddressableRemoteSize * percentageThreshold / 100.0);
+        }
+
+        // Check for absolute bytes threshold
+        final double dataToFileCacheSizeRatio = fileCacheSettings.getRemoteDataRatio();
+        ByteSizeValue bytesThreshold = diskThresholdSettings.getFreeBytesThresholdHigh();
+        if (bytesThreshold != null && bytesThreshold.getBytes() > 0) {
+            return bytesThreshold.getBytes() * (long) dataToFileCacheSizeRatio;
+        }
+
+        // Default fallback
+        return 0;
+    }
+
+    private long calculateCurrentNodeRemoteShardSize(RoutingNode node, RoutingAllocation allocation, boolean subtractLeavingShards) {
+        final List<ShardRouting> remoteShardsOnNode = StreamSupport.stream(node.spliterator(), false)
+            .filter(
+                shard -> shard.primary()
+                    && REMOTE_CAPABLE.equals(getShardPool(shard, allocation))
+                    && (!subtractLeavingShards || !shard.relocating())
+            )
+            .collect(Collectors.toList());
+
+        var remoteShardSize = 0L;
+        for (ShardRouting shard : remoteShardsOnNode) {
+            remoteShardSize += DiskThresholdDecider.getExpectedShardSize(
+                shard,
+                0L,
+                allocation.clusterInfo(),
+                allocation.snapshotShardSizeInfo(),
+                allocation.metadata(),
+                allocation.routingTable()
+            );
+        }
+
+        return remoteShardSize;
+    }
+
+    private long calculateTotalAddressableRemoteSize(RoutingNode node, RoutingAllocation allocation) {
+        ClusterInfo clusterInfo = allocation.clusterInfo();
+        // TODO: Change the default value to 5 instead of 0
+        final double dataToFileCacheSizeRatio = fileCacheSettings.getRemoteDataRatio();
+        final FileCacheStats fileCacheStats = clusterInfo.getNodeFileCacheStats().getOrDefault(node.nodeId(), null);
+        final long nodeCacheSize = fileCacheStats != null ? fileCacheStats.getTotal().getBytes() : 0;
+        return (long) dataToFileCacheSizeRatio * nodeCacheSize;
+    }
+
+    private Decision earlyTerminate(RoutingNode node, RoutingAllocation allocation) {
+        // Always allow allocation if the decider is disabled
+        if (diskThresholdSettings.isEnabled() == false) {
```

Comment:
lets add another setting for disabling warm decider on lines of `cluster.routing.allocation.disk.threshold_enabled`